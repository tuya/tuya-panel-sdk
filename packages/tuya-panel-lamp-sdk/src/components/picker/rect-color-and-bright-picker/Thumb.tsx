import React, { Component } from 'react';
import { View, Image, StyleSheet, Animated, NativeComponent } from 'react-native';

interface IProps {
  img?: any;
  color: string;
  size?: number;
  x: number;
  y: number;
  disabled?: boolean;
}

export default class Thumb extends Component<IProps> {
  componentWillReceiveProps(nextProps: IProps) {
    this.animateX.setValue(nextProps.x);
    this.animateY.setValue(nextProps.y);
  }

  setNativeProps(props: IProps) {
    const { x, y, color } = props;
    if (typeof x === 'number') {
      this.animateX.setValue(x);
    }
    if (typeof y === 'number') {
      this.animateY.setValue(y);
    }
    if (typeof color === 'string') {
      this.previewRef.setNativeProps({
        style: {
          backgroundColor: color,
        },
      });
    }
  }

  private animateX = new Animated.Value(this.props.x);
  private animateY = new Animated.Value(this.props.y);
  private previewRef: NativeComponent;

  render() {
    const { color, size, img } = this.props;
    const maskWidth = (size / 38) * 62;
    const halfWidth = maskWidth / 2;
    return (
      <Animated.View
        style={[
          styles.thumb,
          {
            top: -halfWidth,
            left: -halfWidth,
            width: maskWidth,
            height: maskWidth,
            transform: [{ translateX: this.animateX }, { translateY: this.animateY }],
          },
        ]}
      >
        <View
          ref={(ref: NativeComponent) => {
            this.previewRef = ref;
          }}
          style={{ borderRadius: size / 2, width: size, height: size, backgroundColor: color }}
        />
        <Image
          source={img}
          style={[StyleSheet.absoluteFill, { width: maskWidth, height: maskWidth }]}
        />
      </Animated.View>
    );
  }
}

const styles = StyleSheet.create({
  thumb: {
    alignItems: 'center',
    justifyContent: 'center',
    position: 'absolute',
  },
});
