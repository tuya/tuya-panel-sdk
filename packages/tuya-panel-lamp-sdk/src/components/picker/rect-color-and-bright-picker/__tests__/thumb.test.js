/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow, mount } from 'enzyme';
import Thumb from '../Thumb';

describe('Thumb', () => {
  const origConsole = console.error;
  beforeEach(() => {
    console.error = () => {};
  });
  afterEach(() => {
    console.error = origConsole;
  });

  it('base render', () => {
    const wrapper = shallow(<Thumb x={10} y={0} />);

    wrapper.instance().setNativeProps({ x: 30, y: 20 });

    wrapper.setProps({ x: 10, y: 10 });

    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
  });

  it('basic mount', () => {
    const wrapper = mount(<Thumb x={10} y={0} />);
    expect(wrapper).toMatchSnapshot();

    wrapper.instance().setNativeProps({ color: '#fff' });

    wrapper.unmount();
  });
});
