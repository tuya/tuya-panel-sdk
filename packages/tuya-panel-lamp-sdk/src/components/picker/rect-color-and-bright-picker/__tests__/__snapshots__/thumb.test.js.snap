// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thumb base render 1`] = `
<AnimatedComponent
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
      },
      Object {
        "height": NaN,
        "left": NaN,
        "top": NaN,
        "transform": Array [
          Object {
            "translateX": 10,
          },
          Object {
            "translateY": 10,
          },
        ],
        "width": NaN,
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": undefined,
        "borderRadius": NaN,
        "height": undefined,
        "width": undefined,
      }
    }
  />
  <Image
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": NaN,
          "width": NaN,
        },
      ]
    }
  />
</AnimatedComponent>
`;

exports[`Thumb basic mount 1`] = `
<Thumb
  x={10}
  y={0}
>
  <AnimatedComponent
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "position": "absolute",
        },
        Object {
          "height": NaN,
          "left": NaN,
          "top": NaN,
          "transform": Array [
            Object {
              "translateX": 10,
            },
            Object {
              "translateY": 0,
            },
          ],
          "width": NaN,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "height": NaN,
          "justifyContent": "center",
          "left": NaN,
          "position": "absolute",
          "top": NaN,
          "transform": Array [
            Object {
              "translateX": 10,
            },
            Object {
              "translateY": 0,
            },
          ],
          "width": NaN,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "height": NaN,
            "justifyContent": "center",
            "left": NaN,
            "position": "absolute",
            "top": NaN,
            "transform": Array [
              Object {
                "translateX": 10,
              },
              Object {
                "translateY": 0,
              },
            ],
            "width": NaN,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": undefined,
              "borderRadius": NaN,
              "height": undefined,
              "width": undefined,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": undefined,
                "borderRadius": NaN,
                "height": undefined,
                "width": undefined,
              }
            }
          />
        </View>
        <Image
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              Object {
                "height": NaN,
                "width": NaN,
              },
            ]
          }
        >
          <Image
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {
                  "height": NaN,
                  "width": NaN,
                },
              ]
            }
          />
        </Image>
      </View>
    </View>
  </AnimatedComponent>
</Thumb>
`;
